flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -160,38 +162,58 @@ private void cancelLogicalSlotRequest(ExecutionVertexID executionVertexId, Throw
 			ExecutionSlotSharingGroup group = entry.getKey();
 			List<ExecutionVertexID> executionIds = entry.getValue();
 			SharedSlot sharedSlot = getOrAllocateSharedSlot(group, sharedSlotProfileRetriever);
-
-			for (ExecutionVertexID executionId : executionIds) {
-				CompletableFuture<LogicalSlot> logicalSlotFuture = sharedSlot.allocateLogicalSlot(executionId);
-				SlotExecutionVertexAssignment assignment = new SlotExecutionVertexAssignment(executionId, logicalSlotFuture);
-				assignments.put(executionId, assignment);
-			}
+			allocateLogicalSlotsFromSharedSlot(assignments, executionIds, sharedSlot);
 		}
 
 		return assignments;
 	}
 
 	private SharedSlot getOrAllocateSharedSlot(
-			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			ExecutionSlotSharingGroup slotSharingGroup,
 			SharedSlotProfileRetriever sharedSlotProfileRetriever) {
-		return sharedSlots
-			.computeIfAbsent(executionSlotSharingGroup, group -> {
-				SlotRequestId physicalSlotRequestId = new SlotRequestId();
-				ResourceProfile physicalSlotResourceProfile = getPhysicalSlotResourceProfile(group);
-				SlotProfile slotProfile = sharedSlotProfileRetriever.getSlotProfile(group, physicalSlotResourceProfile);
-				PhysicalSlotRequest physicalSlotRequest =
-					new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely);
-				CompletableFuture<PhysicalSlot> physicalSlotFuture = slotProvider
+		SharedSlot sharedSlot = sharedSlots.get(slotSharingGroup);
+		if (sharedSlot == null) {
+			SlotRequestId physicalSlotRequestId = new SlotRequestId();
+			ResourceProfile physicalSlotResourceProfile =
+					getPhysicalSlotResourceProfile(slotSharingGroup);
+			SlotProfile slotProfile = sharedSlotProfileRetriever
+					.getSlotProfile(slotSharingGroup, physicalSlotResourceProfile);
+			PhysicalSlotRequest physicalSlotRequest = new PhysicalSlotRequest(
+					physicalSlotRequestId,
+					slotProfile,
+					slotWillBeOccupiedIndefinitely);
+			CompletableFuture<PhysicalSlot> physicalSlotFuture = slotProvider
 					.allocatePhysicalSlot(physicalSlotRequest)
 					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
-				return new SharedSlot(
+			sharedSlot = new SharedSlot(
 					physicalSlotRequestId,
 					physicalSlotResourceProfile,
-					group,
+					slotSharingGroup,
 					physicalSlotFuture,
 					slotWillBeOccupiedIndefinitely,
 					this::releaseSharedSlot);
-			});
+			if (!physicalSlotFuture.isCompletedExceptionally()) {
+				sharedSlots.put(slotSharingGroup, sharedSlot);
+			}
```

Comment:
Hmm, this looks a bit like special case handling. I guess it would be bit nicer if the case that it is completed exceptionally now and at a later point would be handled the same way. What exactly is different from a `physicalSlotFuture` being completed exceptionally right away and the `phyiscalSlotFuture` being completed exceptionally just after the `SharedSlot` has been created?